### PR TITLE
fix(griffin) - speed up sym != 

### PR DESCRIPTION
**Pull request by jaugsburger**

### CODE REVIEW

File: core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java [hunk 1]
```diff
@@ -2095,62 +2096,87 @@ private RecordCursorFactory generateTableQuery(
...
-                    symbolValueList.clear();
-
-                    for (int i = 0, n = intrinsicModel.keyValues.size(); i < n; i++) {
-                        symbolValueList.add(intrinsicModel.keyValues.get(i));
-                    }
+                        final Function f = compileFilter(intrinsicModel, readerMeta, executionContext);
+                        if (f != null && f.isConstant() && !f.getBool(null)) {
+                            return new EmptyTableRecordCursorFactory(myMeta);
+                        }
+                        return new FilterOnValuesRecordCursorFactory(
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java:2144]: I recently tackled potential memory leak in Functions. Function is Closeable interface.

When 'f' is constant we will be discarding it regardless if it is true or false. 'false' branch is ok, but 'true' branch needs to behave as if there is no filter. In both cases 'f' need to be closed. I usually do something like:

```java
if (f.isConstant()) {
   try (f) {
       if (f.getBool(null)) {
       } else {
       }
   }
}
```

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 2]
```diff
@@ -125,11 +127,36 @@ private boolean analyzeEquals0(AliasTranslator translator, IntrinsicModel model,
                                         model.keyValues.add(value);
                                         model.keyValuePositions.add(b.position);
                                     }
+                                    node.intrinsicValue = IntrinsicModel.TRUE;
                                 } else {
+                                    if (model.keyExcludedValues.contains(value)) {
+                                        if (model.keyExcludedValues.size() > 1) {
+                                            model.keyExcludedValues.remove(value);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:134]: this branch is untested

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 3]
```diff
@@ -125,11 +127,36 @@ private boolean analyzeEquals0(AliasTranslator translator, IntrinsicModel model,
...
+                                            int removedIndex = model.keyExcludedValues.remove(value);
+                                            if (removedIndex > -1) {
+                                                model.keyExcludedValuePositions.removeIndex(index);
+                                            }
+                                        } else {
+                                            model.keyExcludedValues.clear();
+                                            model.keyExcludedValuePositions.clear();
+                                        }
+                                        tempNodes.clear();
+                                        for (int i = 0; i < keyExclNodes.size(); i++) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:144]: the loops should not be calling `size()` on every iteration. Somewhere else you did:

```sql
for (int n = 0, k = keyExclNodes.size(); n < k; n++) {
```

which is much better

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 4]
```diff
@@ -125,11 +127,36 @@ private boolean analyzeEquals0(AliasTranslator translator, IntrinsicModel model,
...
+                                            ExpressionNode expressionNode = keyExclNodes.get(i);
+                                            if (Chars.equals(expressionNode.lhs.token, b.token) || Chars.equals(expressionNode.rhs.token, b.token)) {
+                                                expressionNode.intrinsicValue = IntrinsicModel.TRUE;
+                                                tempNodes.add(expressionNode);
+                                            }
+                                        }
+                                        for (int i = 0; i < tempNodes.size(); i++) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:151]: same as above

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 5]
```diff
@@ -460,21 +493,78 @@ private boolean analyzeNotEquals0(AliasTranslator translator, IntrinsicModel mod
...
-
-                            // check if we are limited by preferred column
-                            if (preferredKeyColumn != null && !Chars.equals(preferredKeyColumn, column)) {
+                            final boolean preferred = Chars.equalsIgnoreCaseNc(preferredKeyColumn, column);
+                            final boolean indexed = m.isColumnIndexed(index);
+                            if (indexed && preferredKeyColumn == null) {
+                                CharSequence value = isNullKeyword(b.token) ? null : unquote(b.token);
+                                if (Chars.equalsIgnoreCaseNc(model.keyColumn, column)) {
+                                    if (model.keyExcludedValues.contains(value)) {
+                                        // when we have "x not in ('a,'b') and x != 'a')" the x='b' can never happen
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:502]: untested

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 6]
```diff
@@ -460,21 +493,78 @@ private boolean analyzeNotEquals0(AliasTranslator translator, IntrinsicModel mod
...
+                                            model.keyExcludedValuePositions.clear();
+                                            model.keyExcludedValues.add(value);
+                                            model.keyExcludedValuePositions.add(b.position);
+                                            return true;
+                                        }
+                                        node.intrinsicValue = IntrinsicModel.TRUE;
+                                    } else {
+                                        if (model.keyValues.contains(value)) {
+                                            if (model.keyValues.size() > 1) {
+                                                int removedIndex = model.keyValues.remove(value);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:515]: untested

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 7]
```diff
@@ -460,21 +493,78 @@ private boolean analyzeNotEquals0(AliasTranslator translator, IntrinsicModel mod
...
+                                    model.keyExcludedValues.clear();
+                                    model.keyExcludedValuePositions.clear();
+                                    model.keyExcludedValues.add(value);
+                                    model.keyExcludedValuePositions.add(b.position);
+                                    for (int n = 0, k = keyExclNodes.size(); n < k; n++) {
+                                        keyExclNodes.getQuick(n).intrinsicValue = IntrinsicModel.UNDEFINED;
+                                    }
+                                    keyExclNodes.clear();
+                                    for (int n = 0, k = keyNodes.size(); n < k; n++) {
+                                        keyNodes.getQuick(n).intrinsicValue = IntrinsicModel.UNDEFINED;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:551]: untested

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 8]
```diff
@@ -548,16 +638,79 @@ private boolean analyzeNotInInterval(IntrinsicModel model, ExpressionNode col, E
...
             case ColumnType.INT:
-                if (m.isColumnIndexed(columnIndex) && (preferredKeyColumn == null || Chars.equals(preferredKeyColumn, columnName))) {
-                    keyExclNodes.add(notNode);
+
+                boolean newColumn = true;
+
+                if (model.keyColumn != null
+                        && (newColumn = !Chars.equals(model.keyColumn, columnName))
+                        && meta.getIndexValueBlockCapacity(columnIndex) <= meta.getIndexValueBlockCapacity(model.keyColumn)) {
+                    return;
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:655]: untested

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 9]
```diff
@@ -548,16 +638,79 @@ private boolean analyzeNotInInterval(IntrinsicModel model, ExpressionNode col, E
...
+
+
+                int i = node.paramCount - 1;
+                tempKeys.clear();
+                tempPos.clear();
+
+                // collect and analyze values of indexed field
+                // if any of values is not an indexed constant - bail out
+                if (i == 1) {
+                    if (node.rhs == null || node.rhs.type != ExpressionNode.CONSTANT) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:666]: untested

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java [hunk 10]
```diff
@@ -548,16 +638,79 @@ private boolean analyzeNotInInterval(IntrinsicModel model, ExpressionNode col, E
...
+                        return;
+                    }
+                    if (tempKeys.add(unquote(node.rhs.token))) {
+                        tempPos.add(node.position);
+                    }
+                } else {
+                    for (i--; i > -1; i--) {
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/WhereClauseParser.java:673]: untested loop

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+        this.maxSymbolNotEqualsCount = maxSymbolNotEqualsCount;
+        final int nKeyValues = keyValues.size();
+        this.keyExcludedValues.addAll(keyValues);
+        this.columnIndex = columnIndex;
+        this.filter = filter;
+        cursorFactories = new ObjList<>(nKeyValues);
+        if (orderByMnemonic == OrderByMnemonic.ORDER_BY_INVARIANT) {
+            this.cursor = new DataFrameRecordCursor(new SequentialRowCursorFactory(cursorFactories), false, filter, columnIndexes);
+        } else {
+            this.cursor = new DataFrameRecordCursor(new HeapRowCursorFactory(cursorFactories), false, filter, columnIndexes);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java:69]: untested

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+    ) {
+        try (TableReader reader = dataFrameCursor.getTableReader()) {
+            if (reader.getSymbolMapReader(columnIndex).size() > maxSymbolNotEqualsCount) {
+                throw ReaderOutOfDateException.INSTANCE;
+            }
+        }
+        this.recalculateIncludedValues(dataFrameCursor.getTableReader());
+        this.cursor.of(dataFrameCursor, executionContext);
+        if (filter != null) {
+            filter.init(this.cursor, executionContext);
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java:122]: untested

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java (hunk 11 above)
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java:69]: done

File: core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java (hunk 12 above)
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/engine/table/FilterOnExcludedValuesRecordCursorFactory.java:122]: done

File: core/src/main/java/io/questdb/griffin/WhereClauseParser.java (hunk 2 above)
Comment (jaugsburger) [reply core/src/main/java/io/questdb/griffin/WhereClauseParser.java:134]: done

**Pull request merged**